File: src/dataStructures/binarySearchTree.js
class Node {
    constructor(value) {
        this.value = value
        this.left = null
        this.right = null
    }
}
    class BinarySearchTree {
        constructor() {
            this.root = null
        }

        insert(value) {
            let newNode = new Node(value)
            if (!this.root) {
                this.root = newNode
                return this
            }
           let current = this.root
           while (current) {
               if (value===current.value) return undefined // this is one way to handle it.  Other way, create a counter for duplicates
               if (current.value<value) { 
                   if (!current.right) {
                       current.right = newNode
                       current = null
                   } else {
                       current = current.right
                   }
               } else {
                   if (!current.left) {
                       current.left = newNode
                       current = null
                   } else {
                       current = current.left
                   }
               }
           }
           return this
        } 

        find(value) {
            if (!this.root) return false 
            let current = this.root
            while (current) {
                if (current.value===value) return true
                if (value > current.value) {
                    if (!current.right) {
                        return false
                    } else {
                        current = current.right
                    }
                } else {
                    if (!current.left) {
                        return false
                    } else {
                        current = current.left
                    }
                }
            }
        }

        // breath first search
        // return all the values in a Binary Search Tree, touching all of them only once

        breadthFirstSearch() {
            let queue = []
            let result = []
            queue.push(this.root)
            while (queue.length != 0) {
                let dequeue = queue.shift()
                result.push(dequeue)
                if (!!dequeue.left) queue.push(dequeue.left)
                if (!!dequeue.right) queue.push(dequeue.right)
            }
            return result
        }

        depthFirstSearchPreOrder() {
            const result = []
            const traverse = (node) => {
                result.push(node.value)
                if (!!node.left) traverse(node.left)
                if (!!node.right) traverse(node.right)
            }
            traverse(this.root)
            return result
        }

        depthFirstSearchPostOrder() {  // returns the tree in reverse order(ish)
            const result = []
            const traverse = (node) => {
                if (!!node.left) traverse(node.left)
                if (!!node.right) traverse(node.right)
                result.push(node.value)
            }
            traverse(this.root)
            return result
        }

        depthFirstSearchInOrder() {  
            const result = []
            const traverse = (node) => {
                if (!!node.left) traverse(node.left)
                result.push(node.value)
                if (!!node.right) traverse(node.right)
            }
            traverse(this.root)
            return result
        }

// go through this and see how works properly
        remove(value){
            this.root = this.removeNode(this.root, value)
        }
        // a recursive function to insert a new value in binary search tree
        
        removeNode(current, value) {
            
            // base case, if the tree is empty 
            
           if(current === null) return current
            
            // when value is the same as current's value, this is the node to be deleted
            
            if (value === current.value) {
                 
                // for case 1 and 2, node without child or with one child
                
                if (current.left === null && current.right === null){
                    
                        return null
                    
                    }else if(current.left === null){
                    
                        return current.right 
                 
                    }else if(current.right === null){
                    
                        return current.left
                    
                    }else{
                        
                        /// node with two children, get the inorder successor, 
                        //smallest in the right subtree
                        
                        let tempNode = this.kthSmallestNode(current.right)
                            current.value = tempNode.value
                        
                        /// delete the inorder successor
                        
                            current.right = this.removeNode(current.right, tempNode.value)
                        return current
                }
    
            // recur down the tree
                
            }else if(value < current.value) {
                
                current.left = this.removeNode(current.left, value)
                return current
                
            }else{
                
                current.right = this.removeNode(current.right, value)
                return current
            }
        }
        
         /// helper function to find the smallest node
        
        kthSmallestNode(node) {
            while(!node.left === null)
                node = node.left
    
            return node
        }
    
    }

File: src/dataStructures/BinarySearchTree.rb
class Node
    attr_accessor :value, :left, :right

    def initialize(value)
        @value = value
        @left = nil
        @right = nil
    end
end

# BIG O
# Insertion O(log n) = average & best.  Worst is O(n)
# Search O(log n) = average & best.  Worst is O(n)

class BinarySearchTree
    attr_accessor :root

    def initialize
        @root = nil
    end

    def insert(value)
        new_node = Node.new(value)
        if !root
            @root = new_node
            return self
        end
        current = @root
        while current do
            return nil if current.value == value
            if value > current.value
                if !current.right
                    current.right = new_node
                    break
                else 
                    current = current.right
                end
            else 
                if !current.left
                    current.left = new_node
                    break
                else 
                    current = current.left
                end
            end
        end
        return self
    end

    def find(value)
        return false if !@root
        current = @root
        while !!current do
            return true if current.value == value
            if value > current.value
                return false if !current.right
                current = current.right
            else 
                return false if !current.left
                current = current.left
            end
        end
    end

    def breadth_first_search
        queue = []
        result = []
        queue.push(@root)
        while queue.length != 0 do
            dequeued = queue.shift()
            result.push(dequeued)
            queue.push(dequeued.left) if !!dequeued.left
            queue.push(dequeued.right) if !!dequeued.right
        end
        return result
    end

    def depth_first_search_preOrder
        @result = [] # need @ because of scope of helper method
        def travserse(node)
            @result.push(node.value)
            travserse(node.left) if !!node.left
            travserse(node.right) if !!node.right
        end
        travserse(@root)
        return @result 
    end

    def DFS_preorder_memoization(current = self.root, result=[])
        result.push(current.value)
        DFS_preorder_memoization(current.left, result) if !!current.left
        DFS_preorder_memoization(current.right, result) if !!current.right
        return result
    end

    def DFS_postorder_memoization(current = self.root, result = [])
        DFS_postorder_memoization(current.left, result) if !!current.left
        DFS_postorder_memoization(current.right, result) if !!current.right
        result.push(current.value)
        return result
    end


    def depth_first_search_postOrder
        @result = []
        def travserse(node)
            travserse(node.left) if !!node.left
            travserse(node.right) if !!node.right
            @result.push(node.value)
        end
        travserse(@root)
        return @result
    end

    def depth_first_search_in_order
        @result = []
        def travserse(node)
            travserse(node.left) if !!node.left
            @result.push(node.value)
            travserse(node.right) if !!node.right
        end
        travserse(@root)
        return @result
    end


    def find_recursive(value, current = @root)
        return false if !current
        return true if current.value == value
        return find_recursive(value, current.right) if value > current.value
        return find_recursive(value, current.left) if value < current.value
    end

# def delete(value, current = @root)
#     return nil if !current
#     if value < current.value
#         current.left = delete(value, current.left)      
#         elsif 
#             value > current.value
#             current.right = delete(value, current.right)
#         else
#             return current.right if !current.left
#             return current.left if !current.right
#             ## find the maximum value on the left 
#             ## which means, keep travering down the right side
#             right_max = current.left
#             while !!right_max.right do 
#                 right_max = right_max.right
#             end
#             current.left = delete(right_max.value, current.left)
#         end
#     end
#     return true
# end 

def remove(val, current = self.root)
    return nil if !current # tree is empty or we ran out of just_nodes
    if val == current.value # delete this node
      # if delete node has no children
      if !current.left && !current.right
        return nil # meaning, the parent will now point to nil
      elsif # there is only one child of the deleted node
        !current.left
        return current.right
      elsif !current.right
        return current.left
      else # this node has two children, find the smallest node of its right descendents 
        smallest_node = current.right
        while !!smallest_node.left do
          smallest_node = smallest_node.left  
        end 
        current.value = smallest_node.value # move the smallest node of the right side to the spot where we deleted the node, technically, swap its VALUE, not the Node object itself
        current.right = remove(smallest_node.value, current.right) # this line calls this method to delete the node we just found, but at its position that we originally found it, hence, we're calling the "root" of this method the right child of the node we just deleted/replaced
        return current
      end 
    elsif current.value > val
      current.left = remove(val, current.left)
      return current
    else 
      current.right = remove(val, current.right)
      return current
    end 
  end 

end